tests/e2e-playwright/tests/jupyterlabs/jupyterlab.py

```diff
@@ -0,0 +1,142 @@
+
+_WAITING_FOR_SERVICE_TO_START: Final[int] = (
+    10 * MINUTE
+)  # NOTE: smash is 13Gib, math 2Gib
+_WAITING_TIME_FILE_CREATION_PER_GB_IN_TERMINAL: Final[int] = 10 * SECOND
+_SERVICE_NAME_EXPECTED_RESPONSE_TO_WAIT_FOR: Final[dict[str, re.Pattern]] = {
+    "jupyter-math": re.compile(r"/api/contents/workspace"),
+    "jupyter-smash": re.compile(r"/api/contents/workspace"),
+    "jupyter-smash-8-0-0": re.compile(r"/api/contents/workspace"),
+    "jupyter-octave-python-math": re.compile(r"/api/contents"),
+}
+
+_SERVICE_NAME_TAB_TO_WAIT_FOR: Final[dict[str, str]] = {
+    "jupyter-math": "README.ipynb",
+    "jupyter-smash": "README.ipynb",
+    "jupyter-smash-8-0-0": "README.ipynb",
+}
+
+
+@dataclass
+class _JLabTerminalWebSocketWaiter:
+    expected_message_type: Literal["stdout", "stdin"]
+    expected_message_contents: str
+
+    def __call__(self, message: str) -> bool:
+        with log_context(logging.DEBUG, msg=f"handling websocket {message=}"):
+            decoded_message = json.loads(message)
+            if (
+                self.expected_message_type == decoded_message[0]
+                and self.expected_message_contents in decoded_message[1]
+            ):
+                return True
+
+            return False
+
+
+@dataclass
+class _JLabWaitForTerminalWebSocket:
+    def __call__(self, new_websocket: WebSocket) -> bool:
+        with log_context(logging.DEBUG, msg=f"received {new_websocket=}"):
+            if "terminals/websocket" in new_websocket.url:
+                return True
+
+            return False
+
+
+def test_jupyterlab(
+    page: Page,
+    create_project_from_service_dashboard: Callable[
+        [ServiceType, str, str | None], str
+    ],
+    service_key: str,
+    large_file_size: ByteSize,
+    large_file_block_size: ByteSize,
+):
+    # NOTE: this waits for the jupyter to send message, but is not quite enough
+    with log_context(
+        logging.INFO, f"Waiting for {service_key} responsive"
+    ), page.expect_response(
+        _SERVICE_NAME_EXPECTED_RESPONSE_TO_WAIT_FOR[service_key],
+        timeout=_WAITING_FOR_SERVICE_TO_START,
+    ):
+        create_project_from_service_dashboard(ServiceType.DYNAMIC, service_key, None)
+
+    iframe = page.frame_locator("iframe")
+    if service_key == "jupyter-octave-python-math":
+        print(
+            f"skipping any more complicated stuff since this is {service_key=} which is legacy"
+        )
+        return
+
+    with log_context(
+        logging.INFO,
+        f"Waiting for {_SERVICE_NAME_TAB_TO_WAIT_FOR[service_key]} to become visible",
+    ):
+        iframe.get_by_role(
+            "tab", name=_SERVICE_NAME_TAB_TO_WAIT_FOR[service_key]
+        ).wait_for(state="visible")
+    if large_file_size:
+        with log_context(
+            logging.INFO,
+            f"Creating multiple files and 1 file of about {large_file_size.human_readable()}",
+        ):
+            iframe.get_by_role("button", name="New Launcher").click()
```

Comment:
THOUGHT: the structure of this test is great. I wonder if we could use this as a skeleton and inject here a tests that is provided by the service itself.